nistats/permutation_tests.py

```diff
@@ -0,0 +1,296 @@
+        If True, performs an unsigned t-test. Both positive and negative
+        effects are considered; the null hypothesis is that the effect is zero.
+        If False, only positive effects are considered as relevant. The null
+        hypothesis is that the effect is zero or negative.
+
+    random_state : int or None,
+        Seed for random number generator, to have the same permutations
+        in each computing units.
+
+    Returns
+    -------
+    unc_ranks_parts: array-like, shape=(n_voxels)
+        Accumulated count of permuted stats over original stats,
+        to compute uncorrected p-values. (limited to this permutation chunk).
+
+    smax_parts : array-like, shape=(n_perm_chunk, )
+        Distribution of the (max) t-statistic under the null hypothesis
+        (limited to this permutation chunk).
+
+    References
+    ----------
+    [1] Nichols, T. E., & Holmes, A. P. (2002). Nonparametric permutation
+    tests for functional neuroimaging: a primer with examples. Human brain
+    mapping, 15(1), 1-25..
+    """
+    # initialize the seed of the random generator
+    rng = check_random_state(random_state)
+
+    # run the permutations
+    smax_parts = np.empty((n_perm_chunk))
+    unc_rank_parts = np.zeros(len(original_stat))
+    # Infer conditions to permute
+    conditions = np.where(con_val != 0.0)
+    t0 = time.time()
+    for i in range(n_perm_chunk):
+        # Shuffle labels of the conditions in the design matrix
+        design_matrices = []
+
+        for frame_times, paradigm, kwargs in design_matrix_objs:
+            perm_paradigm = paradigm.copy()
+            con_idx = np.where(con_val != 0.0)[0]
+            conditions = [con_labels[x] for x in con_idx]
+            selection = np.array([False for x in range(len(paradigm))])
+            for condition in conditions:
+                selection = selection | (perm_paradigm['name'] == condition)
+            sel_labels = perm_paradigm[selection]
+            labels_idx = sel_labels.index.tolist()
+            perm_labels = rng.permutation(sel_labels['name'].tolist())
+            for lidx, perm_label in zip(labels_idx, perm_labels):
+                perm_paradigm.loc[lidx, 'name'] = perm_label
+
+            design_matrices.append(make_design_matrix(frame_times,
+                                                      perm_paradigm,
+                                                      **kwargs))
+
+        # Get permuted stat
+        glm_obj = clone(glm_ref)
+        permuted_stat, = glm_obj.fit(imgs, design_matrices) \
+                                .transform(con_val, contrast_name=con_id,
+                                           output_z=False, output_stat=True)
+        permuted_stat = glm_obj.masker_.transform(permuted_stat)[0]
+
+        # For uncorrected p-values
+        if two_sided_test:
+            unc_rank_parts += (np.fabs(permuted_stat) < original_stat)
+        else:
+            unc_rank_parts += (permuted_stat < original_stat)
```

Comment:
I thinks that storing the permutation values instead of the result of the threshold would avoid requiring the original stat in this function and would make it easier to understand: it would simply be a generator of statistical values to compare with an unknwon reference. The test would be handled in another function.
Try to keep all the functions as specific and simple as possible.